src/jrd/tpc.cpp

```diff
@@ -146,24 +149,38 @@ void TipCache::finalizeTpc(thread_db* tdbb)
 		} while (m_blocks_memory.getNext());
 	}
 
+	PathName nmSnap, nmHdr;
 	if (m_snapshots)
 	{
-		m_snapshots->removeMapFile();
+		nmSnap = m_snapshots->getMapFileName();
 		delete m_snapshots;
 		m_snapshots = NULL;
 	}
 
 	if (m_tpcHeader)
 	{
-		m_tpcHeader->removeMapFile();
+		nmHdr = m_tpcHeader->getMapFileName();
 		delete m_tpcHeader;
 		m_tpcHeader = NULL;
 	}
 
 	m_blocks_memory.clear();
 	m_transactionsPerBlock = 0;
 
-	LCK_release(tdbb, &lock);
+    if (nmSnap.hasData() || nmHdr.hasData())
+    {
+    	if (LCK_lock(tdbb, m_lock, LCK_EX, LCK_NO_WAIT))
+		{
+			if (nmSnap.hasData())
+				SharedMemoryBase::unlinkFile(nmSnap.c_str());
+			if (nmHdr.hasData())
+				SharedMemoryBase::unlinkFile(nmHdr.c_str());
+		}
+		else
+			tdbb->tdbb_status_vector->init();
+	}
+
+	LCK_release(tdbb, m_lock);
```

Comment:
If LCK_lock() in line 172 failed, here we'll get bugcheck in LM

Reply:
Actually, not a bugcheck (due to additional check in LCK), but anyway - this is not fully correct